packages/world/test/query.t.sol

```diff
@@ -0,0 +1,198 @@
+    assertEq(keyTuplesHasHas.length, 1);
+    assertEq(keyTuplesHasHas[0][0], keys[0][0]);
+
+    // Insert the second key into table A
+    world.setRecord(namespace, nameA, keys[1], abi.encodePacked(value));
+
+    keyTuplesHasNot = query(world, fragmentsHasNot);
+    keyTuplesHasHas = query(world, fragmentsHasHas);
+
+    assertEq(keyTuplesHasNot.length, 0);
+    assertEq(keyTuplesHasHas.length, 2);
+    for (uint256 i; i < 2; i++) {
+      assertEq(keyTuplesHasHas[i][0], keys[i][0]);
+    }
+  }
+
+  function testHasValue() public {
+    vm.assume(value != 0);
+
+    _installKeysInTableModule();
+    _installKeysWithValueModule();
+
+    world.setRecord(namespace, nameA, keys[0], abi.encodePacked(value));
+    world.setRecord(namespace, nameB, keys[0], abi.encodePacked(value));
+    world.setRecord(namespace, nameB, keys[1], abi.encodePacked(value));
+
+    QueryFragment[] memory fragmentsHasHasvalue = new QueryFragment[](2);
+    QueryFragment[] memory fragmentsHasNotvalue = new QueryFragment[](2);
+    QueryFragment[] memory fragmentsHasvalueHas = new QueryFragment[](2);
+    QueryFragment[] memory fragmentsHasvalueHasValue = new QueryFragment[](2);
+    fragmentsHasHasvalue[0] = QueryFragment(QueryType.Has, tableA, "");
+    fragmentsHasHasvalue[1] = QueryFragment(QueryType.HasValue, tableB, abi.encodePacked(value));
+    fragmentsHasNotvalue[0] = QueryFragment(QueryType.Has, tableA, "");
+    fragmentsHasNotvalue[1] = QueryFragment(QueryType.NotValue, tableB, abi.encodePacked(value));
+    fragmentsHasvalueHas[0] = QueryFragment(QueryType.HasValue, tableA, abi.encodePacked(value));
+    fragmentsHasvalueHas[1] = QueryFragment(QueryType.Has, tableB, "");
+    fragmentsHasvalueHasValue[0] = QueryFragment(QueryType.HasValue, tableA, abi.encodePacked(value));
+    fragmentsHasvalueHasValue[1] = QueryFragment(QueryType.HasValue, tableB, abi.encodePacked(value));
+
+    // Assert that all the fragments return the correct keys
+    bytes32[][] memory keyTuplesHasHas = query(world, fragmentsHasHas);
+    assertEq(keyTuplesHasHas.length, 1);
+    assertEq(keyTuplesHasHas[0][0], keys[0][0]);
+    bytes32[][] memory keyTuplesHasNot = query(world, fragmentsHasNot);
+    assertEq(keyTuplesHasNot.length, 0);
+    // !gasreport running a Has, HasValue query with four keys
+    bytes32[][] memory keyTuplesHasHasValue = query(world, fragmentsHasHasvalue);
+    assertEq(keyTuplesHasHasValue.length, 1);
+    assertEq(keyTuplesHasHasValue[0][0], keys[0][0]);
+    bytes32[][] memory keyTuplesHasNotvalue = query(world, fragmentsHasNotvalue);
+    assertEq(keyTuplesHasNotvalue.length, 0);
+    bytes32[][] memory keyTuplesHasvalueHas = query(world, fragmentsHasvalueHas);
+    assertEq(keyTuplesHasvalueHas.length, 1);
+    // !gasreport running a HasValue, HasValue query with four keys
+    bytes32[][] memory keyTuplesHasvalueHasvalue = query(world, fragmentsHasvalueHasValue);
+    assertEq(keyTuplesHasvalueHasvalue.length, 1);
+  }
+}
```

Comment:
would be nice to have some gas tests for much larger arrays (e.g. 100, 1k, 10k values) just to see how these all scale

(I assume they scale horribly because solidity and arrays are not friends)

Reply:
great idea, thanks

Reply:
added some now. we can't test 10k values until #804